pkg/controllers/dashboard/helm/repo.go

```diff
@@ -227,26 +233,42 @@ func (r *repoHandler) download(repoSpec *catalog.RepoSpec, status catalog.RepoSt
 	}
 
 	downloadTime := metav1.Now()
-	if repoSpec.GitRepo != "" && status.IndexConfigMapName == "" {
-		commit, err = git.Head(secret, metadata.Namespace, metadata.Name, repoSpec.GitRepo, repoSpec.GitBranch, repoSpec.InsecureSkipTLSverify, repoSpec.CABundle)
-		if err != nil {
-			return status, err
-		}
-		status.URL = repoSpec.GitRepo
-		status.Branch = repoSpec.GitBranch
-		index, err = git.BuildOrGetIndex(metadata.Namespace, metadata.Name, repoSpec.GitRepo)
-	} else if repoSpec.GitRepo != "" {
-		commit, err = git.Update(secret, metadata.Namespace, metadata.Name, repoSpec.GitRepo, repoSpec.GitBranch, repoSpec.InsecureSkipTLSverify, repoSpec.CABundle)
+	// Decide if we need to perform a download operation
+	if repoSpec.GitRepo != "" {
+		repo, err := git.BuildRepoConfig(secret, metadata.Namespace, metadata.Name, repoSpec.GitRepo, repoSpec.InsecureSkipTLSverify, repoSpec.CABundle)
 		if err != nil {
 			return status, err
 		}
-		status.URL = repoSpec.GitRepo
-		status.Branch = repoSpec.GitBranch
-		if status.Commit == commit {
-			status.DownloadTime = downloadTime
-			return status, nil
+		// We need a download operation
+		// if we don't have a Index ConfigMap name, that means, we have not cloned the repository yet.
+		// if we have one, we just need to update it
+		if status.IndexConfigMapName == "" {
+			commit, err = repo.Head(repoSpec.GitBranch)
+			if err != nil {
+				return status, err
+			}
+			status.URL = repoSpec.GitRepo
+			status.Branch = repoSpec.GitBranch
+			index, err = git.BuildOrGetIndex(metadata.Namespace, metadata.Name, repoSpec.GitRepo)
+			if err != nil || index == nil {
+				return status, err
+			}
+		} else {
+			commit, err = repo.CheckUpdate(repoSpec.GitBranch, settings.SystemCatalog.Get())
+			if err != nil {
+				return status, err
+			}
+			status.URL = repoSpec.GitRepo
+			status.Branch = repoSpec.GitBranch
+			if status.Commit == commit {
+				status.DownloadTime = downloadTime
+				return status, nil
+			}
+			index, err = git.BuildOrGetIndex(metadata.Namespace, metadata.Name, repoSpec.GitRepo)
+			if err != nil || index == nil {
```

Comment:
nit: This if is the same as the one in line 279, so it's not needed. That said, I think this is more readable and maybe we could use it in the else if condition too as well as wrapping the error, so it's easier to debug later

Reply:
Yeah, It was failing in the git CI validation without it. 
I agree with you, but it was just saying that the error and index variable were not being used.

Reply:
The condition on line 279 could remain just with `index == nil`, since it's not certain that it won't be nil anyways. The error condition could be done on line 276 as well, and we could wrap the error on line 269 and 254 too. They both call the same method and if we only wrap it on `git.BuildOrGetIndex` it will be harder to debug

Reply:
Fixing